trinity/sync/full/chain.py

```diff
@@ -61,16 +76,59 @@
 ]
 
 
-class FastChainSyncer(BaseHeaderChainSyncer):
+class WaitingPeers:
     """
-    Sync with the Ethereum network by fetching block headers/bodies and storing them in our DB.
-
-    Here, the run() method returns as soon as we complete a sync with the peer that announced the
-    highest TD, at which point we must run the StateDownloader to fetch the state for our chain
-    head.
+    Peers waiting to perform some action. When getting a peer from this queue,
+    prefer the peer with the best throughput for the given command.
     """
-    NO_PEER_RETRY_PAUSE = 5
-    """If no peers are available for downloading the chain data, retry after this many seconds"""
+    _waiting_peers: 'PriorityQueue[SortableTask[ETHPeer]]'
+
+    def __init__(self, response_command_type: Type[Command]) -> None:
+        self._waiting_peers = PriorityQueue()
+        self._response_command_type = response_command_type
+        self._peer_wrapper = SortableTask.orderable_by_func(self._ranked_peer)
+
+    def _ranked_peer(self, peer: ETHPeer) -> float:
+        relevant_throughputs = [
+            exchange.tracker.items_per_second_ema.value
+            for exchange in peer.requests
+            if exchange.response_cmd_type == self._response_command_type
+        ]
+
+        if len(relevant_throughputs) == 0:
+            raise ValidationError(
+                f"Could not find any exchanges on {peer} "
+                f"with response {self._response_command_type!r}"
+            )
+
+        avg_throughput = sum(relevant_throughputs) / len(relevant_throughputs)
+
+        # high throughput peers should pop out of the queue first, so ranked as negative
+        return -1 * avg_throughput
+
+    def put_nowait(self, peer: ETHPeer) -> None:
+        self._waiting_peers.put_nowait(self._peer_wrapper(peer))
+
+    async def get_fastest(self) -> ETHPeer:
+        wrapped_peer = await self._waiting_peers.get()
+        peer = wrapped_peer.original
+
+        # make sure the peer has not gone offline while waiting in the queue
+        while not peer.is_operational:
+            # if so, look for the next best peer
+            wrapped_peer = await self._waiting_peers.get()
+            peer = wrapped_peer.original
```

Comment:
This suffers from the same problem you have with the previous `peer.is_operational` check. I see the following options.

- remove the check entirely and rely on call sites to do the check (not ideal, probably error prone)
- infinite loop or recursive call (not ideal due to uncertain reliability of exit conditions)
- infinite loop or recursive call **with timeout**  (provides certainty around exit conditions and *guarantees* operation peer)

So maybe change to something like:

```python
async def get_fastest(self, timeout=DEFAULT_TIMEOUT):
    start_at = time.perf_counter()
    while True:
        remaining = timeout - (time.perf_counter() - start_at)
        peer = self.cancel_token.cancellable_wait(self._waiting_peers.get(), timeout=remaining)
        if peer.is_operational:
            return peer
```

I'm sure there is a way to do this without the cancel token....

Reply:
> This suffers from the same problem you have with the previous peer.is_operational check. I see the following options.

Hm, I think it's working as intended. Maybe I'll add a comment to clarify, like:

```py
        while not peer.is_operational:
            # if so, look for the next best peer
            wrapped_peer = await self._waiting_peers.get()
            peer = wrapped_peer.original
            # assign next peer to `peer` so that we keep looping until we find an operational one
```

Reply:
Blah, somehow I read the original loop as an `if` statement.  Ignore me.